### ISSUE TITLE
Gradient returns "Nan" values when using triplet loss with similar batches

### ISSUE THREAD

**Issue by Jarboui** [tensorflow/tensorflow#30199]
Hi, 

I am using tf 1.12.0. 
My issue is related to this [one](https://github.com/tensorflow/tensorflow/issues/783). 
It seems that it was closed judging it the benefit are marginal with respect to the effort.
I stumbled on this problem as I was working on my use case. I am not sure how common this is, but I think it deserves a second look. 

I am working on sequence representation learning, and when generating [my triplet loss batches](https://arxiv.org/abs/1703.07737), I end up sometime generating positives that are similar to my anchor. 
This causes the whole gradient to become "Nan".
![Screenshot from 2019-06-27 13-03-17](https://user-images.githubusercontent.com/38140485/60261705-d547f280-98dc-11e9-9e5d-e7b84348ff0a.png)

The suggested work around of replacing nan values with zeros post-computation isn't sufficient in this case because in reality not the whole batch gradient is null.
PS: I am generating the sequences randomly, however with relatively small vocabulary size and horizon, generating similar sequences does happen.

For those having the same issue, in my use case, the work around I opted for is to either condition my samples or add a small random noise. 

However I think implementing the discussed "zero" solution would be really helpful for future work on sequence representation learning. 

Thanks :)

**Comment by gadagashwini** [comment]
@Jarboui Will it be possible to provide us the ipynb file. It will indeed help us to move faster. Thanks!

**Comment by Jarboui** [comment]
@gadagashwini I can't upload the ipynb in the comment section. 
I will past the code bellow for your convenience :) 


import tensorflow as tf
from keras.layers import *
from keras.models import Model
import numpy as np

# I define the representation model
inp_dim = 5
outp_dim = 5
inp = Input(shape=( inp_dim,), name="input")
outp = Dense(outp_dim, activation='tanh', name="simple_layer")(inp)
model = Model(inputs=inp, outputs=outp) 

# I define over here part of my overal loss (note that defining the full loss doesn't change the output.)
anchor = tf.placeholder(tf.float32, [None, inp_dim], name="representation_shift")
positive = tf.placeholder(tf.float32, [None, inp_dim], name="representation_shift")
outp_anchor = model(anchor)
outp_positive = model(positive)
u = tf.math.subtract(outp_anchor, outp_positive)
u_norm = tf.linalg.norm(u, axis=-1) #u_norm is the loss am trying to minimize in this example

#I define here my gradient operations
_OPTIMIZER = tf.train.AdamOptimizer()
grads_and_vars = _OPTIMIZER.compute_gradients(u_norm)

sess = tf.Session()
init_op = tf.global_variables_initializer()
sess.run(init_op)

# I generate a dummy example
batch_size = 10
random_input1 = np.random.rand(batch_size, inp_dim)
random_input2 = np.random.rand(batch_size, inp_dim)
# And I force it to the case where parts of the data is similar (juste the first line actually)
random_input2[0] = random_input1[0] 

sess.run(grads, feed_dict={
    anchor: random_input1,
    positive: random_input2, 
})[0][0]
# This outputs "nan" erroneously 


I hope this could be helpful :)

**Comment by gadagashwini** [comment]
I could reproduce the issue on colab with Tensorflow 1.12.0.

**Comment by alextp** [comment]
The issue here is that u_norm will be zero and the gradient of a zero norm in TF is NaN. You can use the double-where trick to replace zeros (or small values in general) with something valid, and then things should work:

```
u_safe = tf.where(tf.abs(u) > 0.001, u, tf.ones_like(u))
u_norm = tf.linalg.norm(u_safe, axis=-1) #u_norm is the loss am trying to minimize in this example
```
